Overhaul volto customization training (#860)
plone/training: docs/volto_customization/installation.md

## Prerequisites

Before you start working with this training, ensure you have the following prerequisites:

- <a target="_blank" href="https://nodejs.org/en">Node.js LTS (>=20.x)</a> - (<a target="_blank" href="https://6.docs.plone.org/install/install-from-packages.html#nvm">see instructions for installation</a>)
- <a target="_blank" href="https://www.python.org/">Python</a> - See below for specific versions.
- <a target="_blank" href="https://pipx.pypa.io/stable/">pipx</a>
- <a target="_blank" href="https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating">nvm</a>
- <a target="_blank" href="https://www.gnu.org/software/make/">GNU make</a>
- <a target="_blank" href="https://www.docker.com/get-started">Docker</a> (if using the Plone docker images - <a target="_blank" href="https://6.docs.plone.org/install/containers/index.html">see instructions for installation and usage</a>)

The versions of Python that are supported in Volto depend on the version of Plone that you use.

| Plone | Python       | Volto        |
| ----- | ------------ | ------------ |

> | 6.0   | 3.8-3.12     | 17.0 or 18.0 |

Reviewer: @nileshgulia1 this is not accurate, 6.0 is "LTS" so the recommended version is Volto 16.0.0 which is also "LTS".

Author: fixed in [d91612d](https://github.com/plone/training/pull/860/commits/d91612d721ed191d3f0935f6a5758754357f06f7)

## Prerequisites

Before you start working with this training, ensure you have the following prerequisites:

- <a target="_blank" href="https://nodejs.org/en">Node.js LTS (>=20.x)</a> - (<a target="_blank" href="https://6.docs.plone.org/install/install-from-packages.html#nvm">see instructions for installation</a>)
- <a target="_blank" href="https://www.python.org/">Python</a> - See below for specific versions.
- <a target="_blank" href="https://pipx.pypa.io/stable/">pipx</a>
- <a target="_blank" href="https://github.com/nvm-sh/nvm?tab=readme-ov-file#installing-and-updating">nvm</a>
- <a target="_blank" href="https://www.gnu.org/software/make/">GNU make</a>
- <a target="_blank" href="https://www.docker.com/get-started">Docker</a> (if using the Plone docker images - <a target="_blank" href="https://6.docs.plone.org/install/containers/index.html">see instructions for installation and usage</a>)

> 

Reviewer 2: I would suggest that you replace all the above with a link to https://6.docs.plone.org/install/create-project-cookieplone.html#prerequisites-for-installation. This uses Intersphinx and MyST syntax. Then you'll never have to update it again. 😁 

Also it is not good for accessibility to force users to open links in a new browser window, especially for web app developers who know what they are doing while browsing. I personally despise that behavior, and prefer to command-click, right-click and open in new window, or use my mouse's scroll wheel click.

```suggestion
Follow the guide in Plone 6 Documentation, {ref}`create-project-cookieplone-prerequisites-for-installation-label`.
```

The versions of Python that are supported in Volto depend on the version of Plone that you use.

| Plone | Python       | Volto          |
| ----- | ------------ | -------------- |
| 6.1   | 3.10-3.12    | 18.0           |
| 6.0   | 3.8-3.12     | 16.0,17.0,18.0 |

> | 5.2   | 2.7, 3.6-3.8 | 15.0           |

Reviewer 2: Plone 5.2 is no longer supported, so we shouldn't encourage its usage in any way.

```suggestion
```

Depending on the operating system that you are using, some of the following pre-requisites might change.
They assume you have a macOS/Linux machine.

## Bootstrap a new Plone stack

To bootstrap a new Plone project(with both backend and frontend), you can use [Cookieplone](https://github.com/plone/cookieplone).

> You can use pipx to run cookieplone (<a target="_blank" href="https://6.docs.plone.org/install/create-project-cookieplone.html#generate-the-project">see instructions for installation</a>):

Reviewer 2: Cookieplone is capitalized in narrative text. Please check throughout for proper case.

Also this step is not installation, but generation of a project.

Finally, I think the link at this point is a distraction, as your instructions differ slightly. It would be OK at the end of this section as a `seealso`.

```suggestion
You can use pipx to run Cookieplone to generate a project.
```

```shell
pipx run cookieplone project
```

> You will be presented with a series of prompts. You can also specify the addons you want to install along with the project. You can type the name of the project as `tutorial-project` and name of the addon as `volto-teaser-tutorial` when prompted for the addons.

Reviewer 2: Hyphenate "add-on" wherever it occurs in narrative text. The Cookieplone prompt does not hyphenate it. I'll fix that in the template.

```suggestion
You will be presented with a series of prompts.
You can also specify the add-ons you want to install along with the project.
When prompted for {guilabel}`Volto Addon Name`, enter `volto-teaser-tutorial`.
```

This was confusing because you say to change two things here, then later to accept the defaults except for one. I edited the inconsistent sentence.

```shell
[11/17] Volto Addon Name (volto-project-title): volto-teaser-tutorial
```

You can accept the rest of the default values in square brackets (`[default-option]`) by hitting the {kbd}`Enter` key, or enter your preferred values.
For the training, we will use the default values for the rest of the prompts.

## Install the project

To work on your project, you need to install both the frontend and backend.

> Change your current working directory to `tutorial-project`.

Reviewer 2: This is the default value. Replace all instances throughout the training for consistency.

```suggestion
Change your current working directory to `project-title`.
```

```shell

> cd tutorial-project

Reviewer 2: Same

```

To install both the Plone backend and frontend, use the following command.

```shell
make install
```

## Start Plone

Plone 6 has two servers: one for the frontend, and one for the backend.
As such, we need to maintain two active shell sessions, one for each server, to start your Plone site.

### Start Plone backend

In the currently open session, issue the following command.

```shell
make backend-start
```

### Start Plone frontend

Create a second shell session in a new window.
Start the Plone frontend with the following command.

```shell
make frontend-start
```

Open a browser at the following URL to visit your Plone site.

http://localhost:3000

> You can then login with admin/admin at <a target="_blank" href="http://localhost:3000/login">http://localhost:3000/login</a>.

Reviewer 2: Line 98 gets converted to a hyperlink automatically. All URLs do with our MyST configuration. I would remove this link.

Author: done [6e843e8](https://github.com/plone/training/pull/860/commits/6e843e8662610c197ab51afbdecf96bda948efc8)

## Volto addon

Using Cookieplone we should already have a working Volto project with provided addon. You can find the addon in packages/volto-teaser-tutorial.

```{note}

> You might have noticed that we have `volto.config.js` in the root of the project. This is the volto configuration file allowing us to configure Volto and register addons.The addons list points to the addon we just installed. Cookieplone takes care of registering the addon for us.

Reviewer 2: `{file}` renders with semantic HTML to designate it as a file, as opposed to an inline literal. I'd suggest doing that wherever you have files.

Also use care to capitalize Volto as needed.

```suggestion
You might have noticed that we have {file}`volto.config.js` in the root of the project. This is the Volto configuration file allowing us to configure Volto and register add-ons. The add-ons list points to the add-on we just installed. Cookieplone takes care of registering the add-on for us.
```

```

## Workspaces

pnpm workspaces are a way to manage multiple packages in a single repository. Volto is a monorepo, so we use workspaces to manage the Volto project and its addons along with other packages.

We can define workspaces using the `pnpm-workspace.yaml` file in root of our project. This is taken care for us by Cookieplone.

```yaml
packages:
  # all packages in direct subdirs of packages/
  - "core/packages/*"
  - "packages/*"
```

> All the packages in the `packages` directory will be included in the workspace.

Reviewer 2: Directories should also be marked with `{file}` in narrative text.

```suggestion
All the packages in the {file}`packages` directory will be included in the workspace.
```

The dependencies section maps the package names to the workspace. The `workspace:*` specifier tells pnpm to resolve these dependencies from other packages within the same workspace rather than fetching them from the npm registry.

```json
  "dependencies": {
    "@plone/volto": "workspace:*",
    "@plone/registry": "workspace:*",
    "volto-teaser-tutorial": "workspace:*"
  },
```

```{note}

> We don't need to pin specific workspace before we publish/release our project. pnpm takes care of dynamically updating the versions of these packages when you do `pnpm pack` or `pnpm publish`.

Reviewer 2: ```suggestion
We don't need to pin a specific workspace before we publish or release our project. pnpm takes care of dynamically updating the versions of these packages when you do `pnpm pack` or `pnpm publish`.
```

};
```

> ### typescript configuration

Reviewer 2: Its a proper noun. Replace this and all other instances to "TypeScript".

```suggestion
### TypeScript configuration
```
